specs/phase0/fork-choice.md

```diff
@@ -304,13 +362,69 @@ def store_target_checkpoint_state(store: Store, target: Checkpoint) -> None:
 
 ```python
 def update_latest_messages(store: Store, attesting_indices: Sequence[ValidatorIndex], attestation: Attestation) -> None:
-    target = attestation.data.target
+    attestation_slot = attestation.data.slot
     beacon_block_root = attestation.data.beacon_block_root
     for i in attesting_indices:
-        if i not in store.latest_messages or target.epoch > store.latest_messages[i].epoch:
-            store.latest_messages[i] = LatestMessage(epoch=target.epoch, root=beacon_block_root)
+        if i not in store.latest_messages or attestation_slot > store.latest_messages[i].slot:
+            node_key = get_block_slot_key(beacon_block_root, attestation_slot)
+            store.latest_messages[i] = LatestMessage(slot=attestation_slot, root=node_key)
 ```
 
+#### `add_block_slot_node`
+
+```python
+def add_block_slot_node(store: Store, block: BeaconBlock, node_slot: Slot) -> BlockSlotNode:
+    """
+    Adds a ``BlockSlotNode`` corresponding to ``block`` at slot ``node_slot`` in ``store.block_slot_tree`` 
+    and returns the same.
+
+    Algorithm:
+        1. Check if the parent block's node exists at slot ``block.slot - 1``. Add the node if it doesn't exist. 
+        2. Check if this block's node exists at slot ``block.slot``. Add the node if it doesn't exist. 
+        3. If ``node_slot > block.slot``, add virtual block nodes from slot ``block.slot + 1`` till ``node_slot``.
+    """
+    assert block.slot <= node_slot
+
+    # Check if parent block's node exists at block.slot-1
+    # Check for parent block's node only if current block is not genesis/anchor block
+    # NOTE: This only checks if ``block`` is the genesis block. Arbitrary anchor block capability
+    # is not implemented yet.
+    if block.slot > GENESIS_SLOT:
+        parent_block_node_slot = Slot(block.slot - 1)
+        parent_block_node_key = get_block_slot_key(block.parent_root, parent_block_node_slot)
+        # Add the parent block's node if it doesn't exist
+        if parent_block_node_key not in store.block_slot_tree.keys():
+            parent_block = store.blocks[block.parent_root]
+            store.block_slot_tree[parent_block_node_key] = add_block_slot_node(
+                store, parent_block, parent_block_node_slot)
+    else:
+        parent_block_node_key = Root()
+    
+    # Check if this block's node exists at block.slot
+    block_root = hash_tree_root(block)
+    block_node = BlockSlotNode(
+        block_root=block_root, 
+        slot=block.slot, 
+        parent_node=parent_block_node_key, 
+    )
+    block_node_key = get_block_slot_key(block_root, block.slot)
+    store.block_slot_tree[block_node_key] = block_node
+    if node_slot == block.slot:
+        return block_node
+    
+    # Check virtual block's nodes at block.slot+1 till slot
```

Comment:
I think you can entirely remove the "# Check if this block's node exists at block.slot" section and treat the "concrete" node the same as the virtual nodes

```python
parent_node = store.block_slot_tree[parent_block_node_key]
for slot in range(block.slot, node_slot + 1):
    node = BlockSlotNode(
        block_root=block_root, 
        slot=slot, 
        parent_node=parent_node, 
    )
    node_key =  get_block_slot_key(block_root, slot)
    store.block_slot_tree[node_key] = node
    parent_node = node
return node
```